doc/doxygen/src/driver-guide.md

```diff
@@ -0,0 +1,350 @@
+# Guide for writing a device driver in RIOT
+This document describes the requirement, design objectives, and some
+non-function details when writing device drivers in/for RIOT. The term device
+driver in this context includes all 'CPU-external' devices connected to the CPU
+typically via peripherals like SPI, I2C, UART, GPIO, and similar. CPU
+peripherals itself are in RIOT not considered to be device drivers, but
+peripheral or low-level drivers. Typical devices are network devices like
+radios, Ethernet adapters, sensors, and actuators.
+
+
+## General Design Objectives
+Device drivers should be as easy to use as possible. This implies an
+'intialize->ready' paradigm, meaning, that device drivers should be ready to use
+and in operation right after they have been initialized. On top, devices should
+work with physical values where ever possible. So e.g. sensors should return
+already converted values in some physical unit instead of RAW data, so that
+users can work directly with data from different devices directly without having
+to care for device specific conversion.
+
+Additionally towards ease of use, all device drivers in RIOT should provide a
+similar 'look and feel'. They should behave similar concerning things like their
+state after initialization, like their used data representation and so on.
+
+Secondly, all device drivers should be optimized for minimal RAM/ROM usage, as
+RIOT targets (very) constrained devices. This implies, that instead of exposing
+all thinkable functionality, the drivers should focus exporting and implementing
+a device's core functionality, thus covering ~95% of the use cases. Furthermore
+great care should be put into ...(?)
+
+Third, it should always be possible, to handle more than a single device of one
```

Comment:
This considers e.g using `at86rf231` and `at86rf233` at the same time?

Reply:
Sadly yes. But let us not change any content in this PR, and split any discussion on improving the document to their own PRs. In this PR I only want to move it from the wiki to the official doc, without content changes.

Reply:
ok!

Reply:
> Sadly yes.

I meant sadly no, sorry for the confusion. I'm hugely in favor of enabling people to use sub-gigahertz and 2.4 GHz transceivers at the same time, no matter if the happen to be handled by the same driver or not.

@fabian18 is currently working on this, which will have the positive side effect of getting rid of a lot of `#ifdef`s.